owrrm/src/main/java/com/facebook/openwifi/rrm/optimizers/clientsteering/ClientSteeringState.java

```diff
@@ -0,0 +1,111 @@
+/*
+ * Copyright (c) Meta Platforms, Inc. and affiliates.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+package com.facebook.openwifi.rrm.optimizers.clientsteering;
+
+import java.util.Map;
+import java.util.TreeMap;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+
+/** Singleton class to manage global client steering state */
+public final class ClientSteeringState {
+
+	/** Singleton instance */
+	private static final ClientSteeringState singleton =
+		new ClientSteeringState();
+
+	/** Private default constructor (this is a singleton class) */
+	private ClientSteeringState() {};
+
+	/** Get the singleton instance */
+	public static ClientSteeringState getInstance() {
+		return singleton;
+	}
+
+	/**
+	 * Map from AP serial number to radio bssid to time (JVM monotonic time in
+	 * ms) of the latest attempted client steering action. The {@code Long}
+	 * values are never null.
+	 */
+	private final ConcurrentMap<String, Map<String, Long>> apRadioLastAttempt =
+		new ConcurrentHashMap<>();
+
+	/** Reset the state (e.g., for testing) */
+	public final void reset() {
+		apRadioLastAttempt.clear();
+	}
+
+	/**
+	 * Register the time of the latest client steering attempt by the given AP
+	 * and radio.
+	 *
+	 * @param apSerialNumber non-null AP serial number
+	 * @param bssid non-null radio bssid
+	 * @param currentTimeNs current JVM monotonic time (ns)
+	 */
+	final void registerClientSteeringAttempt(
+		String apSerialNumber,
+		String bssid,
+		long currentTimeNs
+	) {
+		Map<String, Long> radioLastAttempt = apRadioLastAttempt
+			.computeIfAbsent(apSerialNumber, k -> new TreeMap<>());
```

Comment:
again, why `TreeMap`? that should not be your default unless you need the keys in sorted order - it's `O(log n)` while `HashMap` is `O(1)`